inputremapper/gui/components/common.py

```diff
@@ -0,0 +1,183 @@
+        label.set_line_wrap(True)
+        label.set_line_wrap_mode(2)
+        # this affeects how many device entries fit next to each other
+        label.set_width_chars(28)
+        label.set_max_width_chars(28)
+
+        box.add(label)
+
+        box.set_margin_top(18)
+        box.set_margin_bottom(18)
+        box.set_homogeneous(True)
+        box.set_spacing(12)
+
+        # self.set_relief(Gtk.ReliefStyle.NONE)
+
+        self.add(box)
+
+        self.show_all()
+
+        self.connect("toggled", self._on_gtk_toggle)
+
+    def _on_gtk_toggle(self):
+        raise NotImplementedError
+
+    def show_active(self, active):
+        """Show the active state without triggering anything."""
+        with HandlerDisabled(self, self._on_gtk_toggle):
+            self.set_active(active)
+
+
+class FlowBoxWrapper:
+    """A wrapper for a flowbox that contains FlowBoxEntry widgets."""
+
+    def __init__(self, flowbox: Gtk.FlowBox):
+        self._gui = flowbox
+
+    def show_active_entry(self, name: Optional[str]):
+        """Activate the togglebutton that matches the name."""
+        for child in self._gui.get_children():
+            flow_box_entry: FlowBoxEntry = child.get_children()[0]
+            flow_box_entry.show_active(flow_box_entry.name == name)
+
+
+class Breadcrumbs:
+    """Writes a breadcrumbs string into a given label."""
+
+    def __init__(
+        self,
+        message_broker: MessageBroker,
+        label: Gtk.Label,
+        show_device_group: bool = False,
+        show_preset: bool = False,
+        show_mapping: bool = False,
+    ):
+        self._message_broker = message_broker
+        self._gui = label
+        self._connect_message_listener()
+
+        self.show_device_group = show_device_group
+        self.show_preset = show_preset
+        self.show_mapping = show_mapping
+
+        self._group_key: str = ""
+        self._preset_name: str = ""
+        self._mapping_name: str = ""
+
+        label.set_max_width_chars(50)
+        label.set_line_wrap(True)
+        label.set_line_wrap_mode(2)
+
+        self._render()
+
+    def _connect_message_listener(self):
+        self._message_broker.subscribe(MessageType.group, self._on_group_changed)
+        self._message_broker.subscribe(MessageType.preset, self._on_preset_changed)
+        self._message_broker.subscribe(MessageType.mapping, self._on_mapping_changed)
+
+    def _on_preset_changed(self, data: PresetData):
+        self._preset_name = data.name or ""
+        self._render()
+
+    def _on_group_changed(self, data: GroupData):
+        self._group_key = data.group_key
+        self._render()
+
+    def _on_mapping_changed(self, mapping: MappingData):
+        if mapping.name:
+            self._mapping_name = mapping.name
+        elif mapping.event_combination != EventCombination.empty_combination():
+            self._mapping_name = mapping.event_combination.beautify()
+        else:
+            self._mapping_name = _("empty mapping")
```

Comment:
```suggestion
    def _on_mapping_changed(self, mapping: MappingData):
        self._mapping_name = mapping.get_name()
```
Can we create a `Mapping.get_name()` function instead of resolving the correct name here?

There is also similar logic in the `gui.components.editor.MappingSelectionLabel.name` property. Which is no longer needed, as we can just expose the private `MappingSelectionLabel._name` attribute.

Reply:
If the circular import is only because of the typing you can do this:

``` python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
   # imports that are not needed during runtime
```

Reply:
Since it works now I would probably just stick with the split file

Reply:
Is this ready to merge now?

Reply:
looks good.